// @neo4j/graphql/dist/graphql/input-objects/generic-operators/IntScalarFilters.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntListFilters = exports.IntScalarFilters = void 0;
const graphql_1 = require("graphql");
exports.IntScalarFilters = new graphql_1.GraphQLInputObjectType({
    name: "IntScalarFilters",
    description: "Int filters",
    fields: {
        eq: { type: graphql_1.GraphQLInt },
        gt: { type: graphql_1.GraphQLInt },
        gte: { type: graphql_1.GraphQLInt },
        in: { type: new graphql_1.GraphQLList(new graphql_1.GraphQLNonNull(graphql_1.GraphQLInt)) },
        lt: { type: graphql_1.GraphQLInt },
        lte: { type: graphql_1.GraphQLInt },
    },
});
exports.IntListFilters = new graphql_1.GraphQLInputObjectType({
    name: "IntListFilters",
    description: "Int list filters",
    fields: {
        eq: { type: new graphql_1.GraphQLList(new graphql_1.GraphQLNonNull(graphql_1.GraphQLInt)) },
        includes: { type: graphql_1.GraphQLInt },
    },
});
