// @neo4j/graphql/dist/schema/validation/custom-rules/utils/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInnerTypeName = getInnerTypeName;
exports.fromValueKind = fromValueKind;
exports.getPrettyName = getPrettyName;
exports.parseArgumentToInt = parseArgumentToInt;
exports.isArrayType = isArrayType;
exports.findArgumentDefinitionNodeByName = findArgumentDefinitionNodeByName;
exports.assertArgumentType = assertArgumentType;
const graphql_1 = require("graphql");
const neo4j = __importStar(require("neo4j-driver"));
const parse_value_node_1 = require("../../../../schema-model/parser/parse-value-node");
function getInnerTypeName(typeNode) {
    if (typeNode.kind === graphql_1.Kind.LIST_TYPE) {
        return getInnerTypeName(typeNode.type);
    }
    if (typeNode.kind === graphql_1.Kind.NON_NULL_TYPE) {
        return getInnerTypeName(typeNode.type);
    }
    // Kind.NAMED_TYPE
    return typeNode.name.value;
}
function fromValueKind(valueNode, enums, expectedType) {
    switch (valueNode.kind) {
        case graphql_1.Kind.STRING:
            return "string";
        case graphql_1.Kind.INT:
            return "int";
        case graphql_1.Kind.FLOAT:
            return "float";
        case graphql_1.Kind.BOOLEAN:
            return "boolean";
        case graphql_1.Kind.ENUM: {
            const enumType = enums?.find((x) => x.name.value === expectedType);
            const enumValue = enumType?.values?.find((value) => value.name.value === valueNode.value);
            if (enumType && enumValue) {
                return enumType.name.value;
            }
            break;
        }
        default:
            // Kind.OBJECT and Kind.VARIABLE remaining
            return;
    }
}
function getPrettyName(typeNode) {
    if (typeNode.kind === graphql_1.Kind.LIST_TYPE) {
        return `[${getPrettyName(typeNode.type)}]`;
    }
    if (typeNode.kind === graphql_1.Kind.NON_NULL_TYPE) {
        return `${getPrettyName(typeNode.type)}!`;
    }
    // Kind.NAMED_TYPE
    return typeNode.name.value;
}
function parseArgumentToInt(arg) {
    if (arg) {
        const parsed = (0, parse_value_node_1.parseValueNode)(arg.value);
        return neo4j.int(parsed);
    }
    return undefined;
}
function isArrayType(traversedDef) {
    return (traversedDef.type.kind === graphql_1.Kind.LIST_TYPE ||
        (traversedDef.type.kind === graphql_1.Kind.NON_NULL_TYPE && traversedDef.type.type.kind === graphql_1.Kind.LIST_TYPE));
}
function findArgumentDefinitionNodeByName(args, name) {
    return args.find((arg) => arg.name === name);
}
function assertArgumentType(argumentNode, inputValueDefinition) {
    const argType = inputValueDefinition.type;
    const argValue = (0, graphql_1.valueFromASTUntyped)(argumentNode.value);
    let isValid = true;
    let errorMsg = "";
    let errorPath = [];
    (0, graphql_1.coerceInputValue)(argValue, argType, (path, _invalidValue, error) => {
        isValid = false;
        errorMsg = error.message;
        errorPath = path;
    });
    return { isValid, errorMsg, errorPath };
}
