// @neo4j/graphql/dist/translate/queryAST/ast/filters/aggregation/AggregationPropertyFilter.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregationPropertyFilter = void 0;
const cypher_builder_1 = __importDefault(require("@neo4j/cypher-builder"));
const InterfaceEntityAdapter_1 = require("../../../../../schema-model/entity/model-adapters/InterfaceEntityAdapter");
const AggregationFilter_1 = require("./AggregationFilter");
class AggregationPropertyFilter extends AggregationFilter_1.AggregationFilter {
    constructor({ attribute, relationship, logicalOperator, comparisonValue, aggregationOperator, attachedTo = "node", isDeprecated, }) {
        super(relationship, isDeprecated, attachedTo);
        this.attribute = attribute;
        this.comparisonValue = comparisonValue;
        this.logicalOperator = logicalOperator;
        this.aggregationOperator = aggregationOperator;
    }
    getChildren() {
        return [];
    }
    getPropertyRefOrAliasesCase(queryASTContext) {
        const implementationsWithAlias = this.getAliasesToResolve();
        if (implementationsWithAlias) {
            return this.generateCaseForAliasedFields(queryASTContext, implementationsWithAlias);
        }
        return this.getPropertyRef(queryASTContext);
    }
    getSubqueryReturnVariable(queryASTContext) {
        const comparisonVar = new cypher_builder_1.default.Variable();
        const property = this.getPropertyRefOrAliasesCase(queryASTContext);
        if (this.aggregationOperator) {
            let propertyExpr = property;
            if (this.attribute.typeHelper.isString()) {
                propertyExpr = cypher_builder_1.default.size(property);
            }
            const aggregateOperation = this.getAggregateOperation(propertyExpr, this.aggregationOperator);
            return this.getOperation(aggregateOperation);
        }
        else {
            let listExpr;
            if (this.logicalOperator !== "EQUAL" && this.attribute.typeHelper.isString()) {
                listExpr = cypher_builder_1.default.collect(cypher_builder_1.default.size(property));
            }
            else {
                listExpr = cypher_builder_1.default.collect(property);
            }
            const comparisonOperation = this.getOperation(comparisonVar);
            return cypher_builder_1.default.any(comparisonVar, listExpr, comparisonOperation);
        }
    }
    getOperation(expr) {
        return this.createBaseOperation({
            operator: this.logicalOperator,
            expr,
            param: new cypher_builder_1.default.Param(this.comparisonValue),
        });
    }
    getAliasesToResolve() {
        if (!this.relationship || !(this.relationship.target instanceof InterfaceEntityAdapter_1.InterfaceEntityAdapter)) {
            return;
        }
        const aliasedImplementationsMap = this.relationship.target.getImplementationToAliasMapWhereAliased(this.attribute);
        if (!aliasedImplementationsMap.length) {
            return;
        }
        return aliasedImplementationsMap;
    }
    generateCaseForAliasedFields(queryASTContext, concreteLabelsToAttributeAlias) {
        if (!queryASTContext.hasTarget()) {
            throw new Error("No parent node found!");
        }
        const aliasesCase = new cypher_builder_1.default.Case();
        for (const [labels, databaseName] of concreteLabelsToAttributeAlias) {
            aliasesCase
                .when(queryASTContext.target.hasLabels(...labels))
                .then(queryASTContext.target.property(databaseName));
        }
        aliasesCase.else(queryASTContext.target.property(this.attribute.databaseName));
        return aliasesCase;
    }
    getPropertyRef(queryASTContext) {
        if (this.attachedTo === "node") {
            if (!queryASTContext.hasTarget()) {
                throw new Error("No parent node found!");
            }
            return queryASTContext.target.property(this.attribute.databaseName);
        }
        else if (this.attachedTo === "relationship" && queryASTContext.relationship) {
            return queryASTContext.relationship.property(this.attribute.databaseName);
        }
        throw new Error("Transpilation error, relationship on filter not available");
    }
    getAggregateOperation(property, aggregationOperator) {
        switch (aggregationOperator) {
            case "AVERAGE":
                return cypher_builder_1.default.avg(property);
            case "MIN":
            case "SHORTEST":
                return cypher_builder_1.default.min(property);
            case "MAX":
            case "LONGEST":
                return cypher_builder_1.default.max(property);
            case "SUM":
                return cypher_builder_1.default.sum(property);
            default:
                throw new Error(`Invalid operator ${aggregationOperator}`);
        }
    }
}
exports.AggregationPropertyFilter = AggregationPropertyFilter;
