// @neo4j/graphql/dist/schema/validation/custom-rules/valid-types/valid-object-type.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidObjectType = ValidObjectType;
const document_validation_error_1 = require("../utils/document-validation-error");
const is_node_type_1 = require("../utils/location-helpers/is-node-type");
function ValidObjectType(context) {
    const interfaceMap = context.interfacesMap;
    const typeMapWithExtensions = context.typeMapWithExtensions;
    if (!interfaceMap || !typeMapWithExtensions) {
        throw new Error("No typeMapWithExtensions found in the validation context");
    }
    return {
        ObjectTypeDefinition(objectType) {
            const { isValid, errorMsg } = (0, document_validation_error_1.assertValid)(() => assertValidType(objectType));
            if (!isValid) {
                context.reportError((0, document_validation_error_1.createGraphQLError)({
                    nodes: [objectType],
                    errorMsg,
                }));
            }
        },
        InterfaceTypeDefinition(interfaceType) {
            const { isValid, errorMsg } = (0, document_validation_error_1.assertValid)(() => {
                assertValidType(interfaceType);
                assertValidNodeInterface(interfaceType, context);
            });
            if (!isValid) {
                context.reportError((0, document_validation_error_1.createGraphQLError)({
                    nodes: [interfaceType],
                    errorMsg,
                }));
            }
        },
    };
}
function assertValidType(type) {
    if (!type.fields || !type.fields.length) {
        throw new document_validation_error_1.DocumentValidationError("Objects and Interfaces must have one or more fields.", []);
    }
    const privateFieldsCount = type.fields.filter((f) => f.directives?.find((d) => d.name.value === "private")).length;
    const fieldsCount = type.fields.length;
    if (privateFieldsCount === fieldsCount) {
        throw new document_validation_error_1.DocumentValidationError("Objects and Interfaces must have one or more fields.", []);
    }
}
function assertValidNodeInterface(interfaceType, context) {
    const interfaceMap = context.interfacesMap;
    const typeMapWithExtensions = context.typeMapWithExtensions;
    if (!interfaceMap || !typeMapWithExtensions) {
        throw new Error("No typeMapWithExtensions found in the validation context");
    }
    const interfaceName = interfaceType.name.value;
    const interfaceConcreteTypes = interfaceMap[interfaceName] ?? [];
    let isNodeInterface = false;
    let hasNonNodeTypes = false;
    for (const concreteType of interfaceConcreteTypes) {
        const isConcreteTypeANode = (0, is_node_type_1.typeIsANodeType)({
            objectTypeDefinitionNode: concreteType,
            typeMapWithExtensions,
        });
        if (isConcreteTypeANode) {
            isNodeInterface = true;
        }
        else {
            hasNonNodeTypes = true;
        }
    }
    if (isNodeInterface && hasNonNodeTypes) {
        throw new document_validation_error_1.DocumentValidationError("Interface needs to be fully implemented by `@node` types.", []);
    }
}
