// @neo4j/graphql/dist/schema/validation/custom-rules/valid-types/valid-field-types.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidFieldTypes = ValidFieldTypes;
const graphql_1 = require("graphql");
const constants_1 = require("../../../../constants");
const document_validation_error_1 = require("../utils/document-validation-error");
const path_parser_1 = require("../utils/path-parser");
const utils_1 = require("../utils/utils");
function verifyRelationshipFieldType({ traversedDef, }) {
    if (traversedDef.kind !== graphql_1.Kind.FIELD_DEFINITION) {
        // delegate
        return;
    }
    const fieldTypeName = (0, utils_1.getInnerTypeName)(traversedDef.type);
    if (constants_1.SCALAR_TYPES.includes(fieldTypeName)) {
        const scalarTypesNotPermittedErrorMessage = `Invalid field type: Scalar types cannot be relationship targets. Please use an Object type instead.`;
        throw new document_validation_error_1.DocumentValidationError(scalarTypesNotPermittedErrorMessage, []);
    }
    const listTypesErrorMessage = `Invalid field type: List type relationship fields must be non-nullable and have non-nullable entries, please change type to [${fieldTypeName}!]!`;
    const listIsNotNonNullable = traversedDef.type.kind === graphql_1.Kind.LIST_TYPE;
    if (listIsNotNonNullable) {
        throw new document_validation_error_1.DocumentValidationError(listTypesErrorMessage, []);
    }
    if (traversedDef.type.kind === graphql_1.Kind.NON_NULL_TYPE) {
        const ifListThenHasNonNullableEntries = traversedDef.type.type.kind === graphql_1.Kind.LIST_TYPE
            ? traversedDef.type.type.type.kind === graphql_1.Kind.NON_NULL_TYPE
            : true;
        if (!ifListThenHasNonNullableEntries) {
            throw new document_validation_error_1.DocumentValidationError(listTypesErrorMessage, []);
        }
    }
}
function getValidationFunction(directiveName) {
    switch (directiveName) {
        case "relationship":
            return verifyRelationshipFieldType;
        case "declareRelationship":
            return verifyRelationshipFieldType;
        default:
            return;
    }
}
function ValidFieldTypes(context) {
    return {
        Directive(directiveNode, _key, _parent, path, ancestors) {
            const [pathToNode, traversedDef, parentOfTraversedDef] = (0, path_parser_1.getPathToNode)(path, ancestors);
            const validationFn = getValidationFunction(directiveNode.name.value);
            if (!validationFn) {
                return;
            }
            if (!traversedDef) {
                console.error("No last definition traversed");
                return;
            }
            const { isValid, errorMsg, errorPath } = (0, document_validation_error_1.assertValid)(() => validationFn({
                directiveNode,
                traversedDef,
                parentDef: parentOfTraversedDef,
            }));
            if (!isValid) {
                context.reportError((0, document_validation_error_1.createGraphQLError)({
                    nodes: [traversedDef],
                    path: [...pathToNode, ...errorPath],
                    errorMsg,
                }));
            }
        },
    };
}
