// @neo4j/graphql/dist/translate/queryAST/ast/fields/attribute-fields/AttributeField.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttributeField = void 0;
const coalesce_if_needed_1 = require("../../filters/utils/coalesce-if-needed");
const Field_1 = require("../Field");
class AttributeField extends Field_1.Field {
    constructor({ alias, attribute }) {
        super(alias);
        this.attribute = attribute;
    }
    getChildren() {
        return [];
    }
    getFieldName() {
        return this.attribute.name;
    }
    getCypherExpr(target) {
        return target.property(this.attribute.databaseName);
    }
    getProjectionField(variable) {
        const variableProperty = variable.property(this.attribute.databaseName);
        return this.createAttributeProperty(variableProperty);
    }
    createAttributeProperty(variableProperty) {
        if (this.alias !== this.attribute.databaseName || this.attribute.annotations.coalesce) {
            return { [this.alias]: (0, coalesce_if_needed_1.coalesceValueIfNeeded)(this.attribute, variableProperty) };
        }
        return this.attribute.databaseName;
    }
}
exports.AttributeField = AttributeField;
