// @neo4j/graphql/dist/translate/queryAST/ast/filters/property-filters/DateTimeFilter.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateTimeFilter = void 0;
const cypher_builder_1 = __importDefault(require("@neo4j/cypher-builder"));
const create_date_time_operation_1 = require("../utils/create-date-time-operation");
const PropertyFilter_1 = require("./PropertyFilter");
class DateTimeFilter extends PropertyFilter_1.PropertyFilter {
    getOperation(prop) {
        return (0, create_date_time_operation_1.createDateTimeOperation)({
            operator: this.operator || "EQ",
            property: prop,
            param: new cypher_builder_1.default.Param(this.comparisonValue),
            attribute: this.attribute,
        });
    }
}
exports.DateTimeFilter = DateTimeFilter;
