// @neo4j/graphql/dist/translate/queryAST/ast/filters/property-filters/PropertyFilter.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PropertyFilter = void 0;
const cypher_builder_1 = __importDefault(require("@neo4j/cypher-builder"));
const InterfaceEntityAdapter_1 = require("../../../../../schema-model/entity/model-adapters/InterfaceEntityAdapter");
const create_comparison_operator_1 = require("../../../utils/create-comparison-operator");
const Filter_1 = require("../Filter");
const coalesce_if_needed_1 = require("../utils/coalesce-if-needed");
class PropertyFilter extends Filter_1.Filter {
    constructor({ attribute, relationship, comparisonValue, operator, attachedTo = "node", }) {
        super();
        this.attribute = attribute;
        this.relationship = relationship;
        this.comparisonValue = comparisonValue;
        this.operator = operator;
        this.attachedTo = attachedTo;
    }
    getChildren() {
        return [];
    }
    print() {
        return `${super.print()} [${this.attribute.name}] <${this.operator}>`;
    }
    getPredicate(queryASTContext) {
        const prop = this.getPropertyRefOrAliasesCase(queryASTContext);
        if (this.comparisonValue === null) {
            return cypher_builder_1.default.isNull(prop);
        }
        return this.getOperation(prop);
    }
    getPropertyRefOrAliasesCase(queryASTContext) {
        const implementationsWithAlias = this.getAliasesToResolve();
        if (implementationsWithAlias) {
            return this.generateCaseForAliasedFields(queryASTContext, implementationsWithAlias);
        }
        return this.getPropertyRef(queryASTContext);
    }
    getAliasesToResolve() {
        if (!this.relationship || !(this.relationship.target instanceof InterfaceEntityAdapter_1.InterfaceEntityAdapter)) {
            return;
        }
        const aliasedImplementationsMap = this.relationship.target.getImplementationToAliasMapWhereAliased(this.attribute);
        if (!aliasedImplementationsMap.length) {
            return;
        }
        return aliasedImplementationsMap;
    }
    generateCaseForAliasedFields(queryASTContext, concreteLabelsToAttributeAlias) {
        if (!queryASTContext.hasTarget()) {
            throw new Error("No parent node found!");
        }
        const aliasesCase = new cypher_builder_1.default.Case();
        for (const [labels, databaseName] of concreteLabelsToAttributeAlias) {
            aliasesCase
                .when(queryASTContext.target.hasLabels(...labels))
                .then(queryASTContext.target.property(databaseName));
        }
        aliasesCase.else(queryASTContext.target.property(this.attribute.databaseName));
        return aliasesCase;
    }
    getPropertyRef(queryASTContext) {
        if (this.attachedTo === "node") {
            if (!queryASTContext.hasTarget()) {
                throw new Error("No parent node found!");
            }
            return queryASTContext.target.property(this.attribute.databaseName);
        }
        else if (this.attachedTo === "relationship" && queryASTContext.relationship) {
            return queryASTContext.relationship.property(this.attribute.databaseName);
        }
        else {
            throw new Error("Transpilation error");
        }
    }
    /** Returns the operation for a given filter.
     * To be overridden by subclasses
     */
    getOperation(prop) {
        return this.createBaseOperation({
            operator: this.operator,
            property: prop,
            param: new cypher_builder_1.default.Param(this.comparisonValue),
        });
    }
    /** Returns the default operation for a given filter */
    createBaseOperation({ operator, property, param, }) {
        const coalesceProperty = (0, coalesce_if_needed_1.coalesceValueIfNeeded)(this.attribute, property);
        return (0, create_comparison_operator_1.createComparisonOperation)({ operator, property: coalesceProperty, param });
    }
}
exports.PropertyFilter = PropertyFilter;
