// @neo4j/graphql/dist/graphql/scalars/DateTime.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLDateTime = void 0;
const graphql_1 = require("graphql");
const neo4j_driver_1 = require("neo4j-driver");
exports.GraphQLDateTime = new graphql_1.GraphQLScalarType({
    name: "DateTime",
    description: "A date and time, represented as an ISO-8601 string",
    serialize: (outputValue) => {
        if (typeof outputValue === "string") {
            return new Date(outputValue).toISOString();
        }
        if ((0, neo4j_driver_1.isDateTime)(outputValue)) {
            return new Date(outputValue.toString()).toISOString();
        }
        throw new graphql_1.GraphQLError(`DateTime cannot represent value: ${outputValue}`);
    },
    parseValue: (inputValue) => {
        if (typeof inputValue === "string") {
            const date = new Date(inputValue);
            if (date.toString() === "Invalid Date") {
                throw new graphql_1.GraphQLError(`DateTime cannot represent non temporal value: ${inputValue}`);
            }
            return inputValue;
        }
        throw new graphql_1.GraphQLError(`DateTime cannot represent non string value: ${inputValue}`);
    },
    parseLiteral(ast) {
        if (ast.kind !== graphql_1.Kind.STRING) {
            throw new graphql_1.GraphQLError("DateTime cannot represent non string value.");
        }
        return ast.value;
    },
});
