// @neo4j/cypher-builder/dist/utils/utils.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.toCypherParams = exports.escapeVariable = exports.escapeType = exports.escapeProperty = exports.escapeLabel = exports.concat = void 0;
// Note: This file exists for exported utils to the user
var concat_1 = require("./concat");
Object.defineProperty(exports, "concat", { enumerable: true, get: function () { return concat_1.concat; } });
var escape_1 = require("./escape");
Object.defineProperty(exports, "escapeLabel", { enumerable: true, get: function () { return escape_1.escapeLabel; } });
Object.defineProperty(exports, "escapeProperty", { enumerable: true, get: function () { return escape_1.escapeProperty; } });
Object.defineProperty(exports, "escapeType", { enumerable: true, get: function () { return escape_1.escapeType; } });
Object.defineProperty(exports, "escapeVariable", { enumerable: true, get: function () { return escape_1.escapeVariable; } });
var to_cypher_params_1 = require("./to-cypher-params");
Object.defineProperty(exports, "toCypherParams", { enumerable: true, get: function () { return to_cypher_params_1.toCypherParams; } });
