// @neo4j/graphql/dist/translate/queryAST/ast/selection/FulltextSelection.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FulltextSelection = void 0;
const cypher_builder_1 = __importDefault(require("@neo4j/cypher-builder"));
const map_labels_with_context_1 = require("../../../../schema-model/utils/map-labels-with-context");
const QueryASTContext_1 = require("../QueryASTContext");
const EntitySelection_1 = require("./EntitySelection");
class FulltextSelection extends EntitySelection_1.EntitySelection {
    constructor({ target, fulltextOptions, scoreVariable, }) {
        super();
        this.target = target;
        this.fulltextOptions = fulltextOptions;
        this.scoreVariable = scoreVariable;
    }
    apply(context) {
        const node = new cypher_builder_1.default.Node();
        const phraseParam = new cypher_builder_1.default.Param(this.fulltextOptions.phrase);
        const indexName = new cypher_builder_1.default.Literal(this.fulltextOptions.index.indexName);
        const fulltextClause = cypher_builder_1.default.db.index.fulltext
            .queryNodes(indexName, phraseParam)
            .yield(["node", node], ["score", this.scoreVariable]);
        const expectedLabels = (0, map_labels_with_context_1.mapLabelsWithContext)(this.target.getLabels(), context.neo4jGraphQLContext);
        const whereOperators = expectedLabels.map((label) => {
            return cypher_builder_1.default.in(new cypher_builder_1.default.Param(label), cypher_builder_1.default.labels(node));
        });
        fulltextClause.where(cypher_builder_1.default.and(...whereOperators));
        return {
            selection: fulltextClause,
            nestedContext: new QueryASTContext_1.QueryASTContext({
                target: node,
                neo4jGraphQLContext: context.neo4jGraphQLContext,
                returnVariable: context.returnVariable,
                env: context.env,
                shouldCollect: context.shouldCollect,
            }),
        };
    }
}
exports.FulltextSelection = FulltextSelection;
