// @neo4j/graphql/dist/translate/queryAST/ast/input-fields/IdField.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdField = void 0;
const cypher_builder_1 = __importDefault(require("@neo4j/cypher-builder"));
const InputField_1 = require("./InputField");
class IdField extends InputField_1.InputField {
    constructor(name, attribute, attachedTo = "node") {
        super(name, attachedTo);
        this.attribute = attribute;
    }
    getChildren() {
        return [];
    }
    print() {
        return `${super.print()} <${this.name}>`;
    }
    getSetParams(queryASTContext) {
        const target = this.getTarget(queryASTContext);
        const setParam = [target.property(this.attribute.databaseName), cypher_builder_1.default.randomUUID()];
        return [setParam];
    }
    getSetClause() {
        return [];
    }
}
exports.IdField = IdField;
