// @neo4j/graphql/dist/translate/queryAST/ast/input-fields/MutationOperationField.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MutationOperationField = void 0;
const InputField_1 = require("./InputField");
class MutationOperationField extends InputField_1.InputField {
    constructor(name, mutationOperation) {
        super(name);
        this.mutationOperation = mutationOperation;
    }
    getChildren() {
        return [this.mutationOperation];
    }
    print() {
        return `${super.print()} <${this.name}>`;
    }
    getSetParams() {
        return [];
    }
    getSubqueries(queryASTContext) {
        const { clauses } = this.mutationOperation.transpile(queryASTContext);
        return clauses;
    }
}
exports.MutationOperationField = MutationOperationField;
