// @neo4j/graphql/dist/graphql/directives/type-dependant-directives/static-definitions.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUBSCRIPTIONS_AUTHORIZATION_FILTER_EVENT = exports.AUTHENTICATION_OPERATION = exports.AUTHORIZATION_FILTER_OPERATION = exports.AUTHORIZATION_VALIDATE_OPERATION = exports.AUTHORIZATION_VALIDATE_STAGE = void 0;
const graphql_1 = require("graphql");
exports.AUTHORIZATION_VALIDATE_STAGE = new graphql_1.GraphQLEnumType({
    name: "AuthorizationValidateStage",
    values: { BEFORE: { value: "BEFORE" }, AFTER: { value: "AFTER" } },
});
exports.AUTHORIZATION_VALIDATE_OPERATION = new graphql_1.GraphQLEnumType({
    name: "AuthorizationValidateOperation",
    values: {
        CREATE: { value: "CREATE" },
        READ: { value: "READ" },
        AGGREGATE: { value: "AGGREGATE" },
        UPDATE: { value: "UPDATE" },
        DELETE: { value: "DELETE" },
        CREATE_RELATIONSHIP: { value: "CREATE_RELATIONSHIP" },
        DELETE_RELATIONSHIP: { value: "DELETE_RELATIONSHIP" },
    },
});
exports.AUTHORIZATION_FILTER_OPERATION = new graphql_1.GraphQLEnumType({
    name: "AuthorizationFilterOperation",
    values: {
        READ: { value: "READ" },
        AGGREGATE: { value: "AGGREGATE" },
        UPDATE: { value: "UPDATE" },
        DELETE: { value: "DELETE" },
        CREATE_RELATIONSHIP: { value: "CREATE_RELATIONSHIP" },
        DELETE_RELATIONSHIP: { value: "DELETE_RELATIONSHIP" },
    },
});
exports.AUTHENTICATION_OPERATION = new graphql_1.GraphQLEnumType({
    name: "AuthenticationOperation",
    values: {
        CREATE: { value: "CREATE" },
        READ: { value: "READ" },
        AGGREGATE: { value: "AGGREGATE" },
        UPDATE: { value: "UPDATE" },
        DELETE: { value: "DELETE" },
        CREATE_RELATIONSHIP: { value: "CREATE_RELATIONSHIP" },
        DELETE_RELATIONSHIP: { value: "DELETE_RELATIONSHIP" },
        SUBSCRIBE: { value: "SUBSCRIBE" },
    },
});
exports.SUBSCRIPTIONS_AUTHORIZATION_FILTER_EVENT = new graphql_1.GraphQLEnumType({
    name: "SubscriptionsAuthorizationFilterEvent",
    values: {
        CREATED: { value: "CREATED" },
        UPDATED: { value: "UPDATED" },
        DELETED: { value: "DELETED" },
    },
});
