// @neo4j/cypher-builder/dist/references/RelationshipRef.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamedRelationship = exports.RelationshipRef = void 0;
const HasLabel_1 = require("../expressions/HasLabel");
const Variable_1 = require("./Variable");
/** Reference to a relationship property
 * @group Variables
 */
class RelationshipRef extends Variable_1.Variable {
    constructor() {
        super();
        this.prefix = "this";
    }
    hasType(label) {
        if (typeof label === "string") {
            return new HasLabel_1.HasLabel(this, [label]);
        }
        else {
            return new HasLabel_1.HasLabel(this, label);
        }
    }
}
exports.RelationshipRef = RelationshipRef;
/** Represents a relationship reference with a given name
 * @group Variables
 */
class NamedRelationship extends RelationshipRef {
    constructor(id) {
        super();
        this.id = id;
    }
    get name() {
        return this.id;
    }
}
exports.NamedRelationship = NamedRelationship;
