// @neo4j/graphql/dist/graphql/input-objects/generic-mutation-operations/BigIntScalarMutations.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BigIntListMutations = exports.BigIntScalarMutations = void 0;
const graphql_1 = require("graphql");
const scalars_1 = require("../../scalars");
const ListMutation_1 = require("./ListMutation");
exports.BigIntScalarMutations = new graphql_1.GraphQLInputObjectType({
    name: "BigIntScalarMutations",
    description: "BigInt mutations",
    fields: {
        set: { type: scalars_1.GraphQLBigInt },
        add: { type: scalars_1.GraphQLBigInt },
        subtract: { type: scalars_1.GraphQLBigInt },
    },
});
exports.BigIntListMutations = (0, ListMutation_1.listMutation)(scalars_1.GraphQLBigInt);
