// @neo4j/graphql/dist/graphql/input-objects/generic-mutation-operations/FloatScalarMutations.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FloatListMutations = exports.FloatScalarMutations = void 0;
const graphql_1 = require("graphql");
const ListMutation_1 = require("./ListMutation");
exports.FloatScalarMutations = new graphql_1.GraphQLInputObjectType({
    name: "FloatScalarMutations",
    description: "Float mutations",
    fields: {
        set: { type: graphql_1.GraphQLFloat },
        add: { type: graphql_1.GraphQLFloat },
        subtract: { type: graphql_1.GraphQLFloat },
        multiply: { type: graphql_1.GraphQLFloat },
        divide: { type: graphql_1.GraphQLFloat },
    },
});
exports.FloatListMutations = (0, ListMutation_1.listMutation)(graphql_1.GraphQLFloat);
