// @neo4j/cypher-builder/dist/Cypher.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MapProjection = exports.Map = exports.PatternComprehension = exports.List = exports.ListComprehension = exports.vector = exports.tx = exports.genai = exports.db = exports.apoc = exports.Exists = exports.Count = exports.Collect = exports.TYPE = exports.isType = exports.isNotType = exports.Case = exports.Variable = exports.NamedVariable = exports.Relationship = exports.NamedRelationship = exports.Property = exports.PathVariable = exports.NamedPathVariable = exports.Param = exports.NamedParam = exports.Node = exports.NamedNode = exports.true = exports.Null = exports.Literal = exports.false = exports.QuantifiedPath = exports.Pattern = exports.labelExpr = exports.With = exports.Use = exports.Unwind = exports.Union = exports.Return = exports.Raw = exports.Merge = exports.OptionalMatch = exports.Match = exports.LoadCSV = exports.Foreach = exports.Finish = exports.Create = exports.OptionalCall = exports.Call = void 0;
exports.degrees = exports.cot = exports.cos = exports.ceil = exports.atan2 = exports.atan = exports.asin = exports.acos = exports.abs = exports.linenumber = exports.file = exports.graph = exports.sum = exports.stDevP = exports.stDev = exports.percentileDisc = exports.percentileCont = exports.min = exports.max = exports.count = exports.collect = exports.avg = exports.Function = exports.pow = exports.plus = exports.multiply = exports.mod = exports.minus = exports.divide = exports.ConcatOp = exports.concat = exports.startsWith = exports.neq = exports.matches = exports.lte = exports.lt = exports.isNull = exports.isNotNull = exports.isNotNormalized = exports.isNormalized = exports.in = exports.gte = exports.gt = exports.eq = exports.endsWith = exports.contains = exports.xor = exports.or = exports.not = exports.and = void 0;
exports.utils = exports.VoidProcedure = exports.Procedure = exports.time = exports.localtime = exports.localdatetime = exports.duration = exports.datetime = exports.date = exports.single = exports.none = exports.isEmpty = exports.exists = exports.any = exports.all = exports.tan = exports.sqrt = exports.sin = exports.sign = exports.round = exports.rand = exports.radians = exports.pi = exports.log10 = exports.log = exports.isNaN = exports.haversin = exports.floor = exports.exp = exports.e = void 0;
// Clauses
var Call_1 = require("./clauses/Call");
Object.defineProperty(exports, "Call", { enumerable: true, get: function () { return Call_1.Call; } });
Object.defineProperty(exports, "OptionalCall", { enumerable: true, get: function () { return Call_1.OptionalCall; } });
var Create_1 = require("./clauses/Create");
Object.defineProperty(exports, "Create", { enumerable: true, get: function () { return Create_1.Create; } });
var Finish_1 = require("./clauses/Finish");
Object.defineProperty(exports, "Finish", { enumerable: true, get: function () { return Finish_1.Finish; } });
var Foreach_1 = require("./clauses/Foreach");
Object.defineProperty(exports, "Foreach", { enumerable: true, get: function () { return Foreach_1.Foreach; } });
var LoadCSV_1 = require("./clauses/LoadCSV");
Object.defineProperty(exports, "LoadCSV", { enumerable: true, get: function () { return LoadCSV_1.LoadCSV; } });
var Match_1 = require("./clauses/Match");
Object.defineProperty(exports, "Match", { enumerable: true, get: function () { return Match_1.Match; } });
Object.defineProperty(exports, "OptionalMatch", { enumerable: true, get: function () { return Match_1.OptionalMatch; } });
var Merge_1 = require("./clauses/Merge");
Object.defineProperty(exports, "Merge", { enumerable: true, get: function () { return Merge_1.Merge; } });
var Raw_1 = require("./clauses/Raw");
Object.defineProperty(exports, "Raw", { enumerable: true, get: function () { return Raw_1.Raw; } });
var Return_1 = require("./clauses/Return");
Object.defineProperty(exports, "Return", { enumerable: true, get: function () { return Return_1.Return; } });
var Union_1 = require("./clauses/Union");
Object.defineProperty(exports, "Union", { enumerable: true, get: function () { return Union_1.Union; } });
var Unwind_1 = require("./clauses/Unwind");
Object.defineProperty(exports, "Unwind", { enumerable: true, get: function () { return Unwind_1.Unwind; } });
var Use_1 = require("./clauses/Use");
Object.defineProperty(exports, "Use", { enumerable: true, get: function () { return Use_1.Use; } });
var With_1 = require("./clauses/With");
Object.defineProperty(exports, "With", { enumerable: true, get: function () { return With_1.With; } });
// Patterns
var label_expressions_1 = require("./expressions/labels/label-expressions");
Object.defineProperty(exports, "labelExpr", { enumerable: true, get: function () { return label_expressions_1.labelExpr; } });
var Pattern_1 = require("./pattern/Pattern");
Object.defineProperty(exports, "Pattern", { enumerable: true, get: function () { return Pattern_1.Pattern; } });
var QuantifiedPath_1 = require("./pattern/quantified-patterns/QuantifiedPath");
Object.defineProperty(exports, "QuantifiedPath", { enumerable: true, get: function () { return QuantifiedPath_1.QuantifiedPath; } });
// Variables and references
var Literal_1 = require("./references/Literal");
Object.defineProperty(exports, "false", { enumerable: true, get: function () { return Literal_1.CypherFalse; } });
Object.defineProperty(exports, "Literal", { enumerable: true, get: function () { return Literal_1.Literal; } });
Object.defineProperty(exports, "Null", { enumerable: true, get: function () { return Literal_1.CypherNull; } });
Object.defineProperty(exports, "true", { enumerable: true, get: function () { return Literal_1.CypherTrue; } });
var NodeRef_1 = require("./references/NodeRef");
Object.defineProperty(exports, "NamedNode", { enumerable: true, get: function () { return NodeRef_1.NamedNode; } });
Object.defineProperty(exports, "Node", { enumerable: true, get: function () { return NodeRef_1.NodeRef; } });
var Param_1 = require("./references/Param");
Object.defineProperty(exports, "NamedParam", { enumerable: true, get: function () { return Param_1.NamedParam; } });
Object.defineProperty(exports, "Param", { enumerable: true, get: function () { return Param_1.Param; } });
var Path_1 = require("./references/Path");
Object.defineProperty(exports, "NamedPathVariable", { enumerable: true, get: function () { return Path_1.NamedPathVariable; } });
Object.defineProperty(exports, "PathVariable", { enumerable: true, get: function () { return Path_1.PathVariable; } });
var PropertyRef_1 = require("./references/PropertyRef");
Object.defineProperty(exports, "Property", { enumerable: true, get: function () { return PropertyRef_1.PropertyRef; } });
var RelationshipRef_1 = require("./references/RelationshipRef");
Object.defineProperty(exports, "NamedRelationship", { enumerable: true, get: function () { return RelationshipRef_1.NamedRelationship; } });
Object.defineProperty(exports, "Relationship", { enumerable: true, get: function () { return RelationshipRef_1.RelationshipRef; } });
var Variable_1 = require("./references/Variable");
Object.defineProperty(exports, "NamedVariable", { enumerable: true, get: function () { return Variable_1.NamedVariable; } });
Object.defineProperty(exports, "Variable", { enumerable: true, get: function () { return Variable_1.Variable; } });
// Expressions
var Case_1 = require("./expressions/Case");
Object.defineProperty(exports, "Case", { enumerable: true, get: function () { return Case_1.Case; } });
var IsType_1 = require("./expressions/IsType");
Object.defineProperty(exports, "isNotType", { enumerable: true, get: function () { return IsType_1.isNotType; } });
Object.defineProperty(exports, "isType", { enumerable: true, get: function () { return IsType_1.isType; } });
Object.defineProperty(exports, "TYPE", { enumerable: true, get: function () { return IsType_1.CypherTypes; } });
// Subquery Expressions
var Collect_1 = require("./expressions/subquery/Collect");
Object.defineProperty(exports, "Collect", { enumerable: true, get: function () { return Collect_1.Collect; } });
var Count_1 = require("./expressions/subquery/Count");
Object.defineProperty(exports, "Count", { enumerable: true, get: function () { return Count_1.Count; } });
var Exists_1 = require("./expressions/subquery/Exists");
Object.defineProperty(exports, "Exists", { enumerable: true, get: function () { return Exists_1.Exists; } });
/**
 * @hideGroups
 */
exports.apoc = __importStar(require("./namespaces/apoc/apoc"));
exports.db = __importStar(require("./namespaces/db/db"));
/**
 * @hideGroups
 */
exports.genai = __importStar(require("./namespaces/genai/genai"));
exports.tx = __importStar(require("./namespaces/tx"));
/**
 * @hideGroups
 */
exports.vector = __importStar(require("./namespaces/vector/vector"));
// --Lists
var ListComprehension_1 = require("./expressions/list/ListComprehension");
Object.defineProperty(exports, "ListComprehension", { enumerable: true, get: function () { return ListComprehension_1.ListComprehension; } });
var ListExpr_1 = require("./expressions/list/ListExpr");
Object.defineProperty(exports, "List", { enumerable: true, get: function () { return ListExpr_1.ListExpr; } });
var PatternComprehension_1 = require("./expressions/list/PatternComprehension");
Object.defineProperty(exports, "PatternComprehension", { enumerable: true, get: function () { return PatternComprehension_1.PatternComprehension; } });
// --Map
var MapExpr_1 = require("./expressions/map/MapExpr");
Object.defineProperty(exports, "Map", { enumerable: true, get: function () { return MapExpr_1.MapExpr; } });
var MapProjection_1 = require("./expressions/map/MapProjection");
Object.defineProperty(exports, "MapProjection", { enumerable: true, get: function () { return MapProjection_1.MapProjection; } });
// --Operations
var boolean_1 = require("./expressions/operations/boolean");
Object.defineProperty(exports, "and", { enumerable: true, get: function () { return boolean_1.and; } });
Object.defineProperty(exports, "not", { enumerable: true, get: function () { return boolean_1.not; } });
Object.defineProperty(exports, "or", { enumerable: true, get: function () { return boolean_1.or; } });
Object.defineProperty(exports, "xor", { enumerable: true, get: function () { return boolean_1.xor; } });
var comparison_1 = require("./expressions/operations/comparison");
Object.defineProperty(exports, "contains", { enumerable: true, get: function () { return comparison_1.contains; } });
Object.defineProperty(exports, "endsWith", { enumerable: true, get: function () { return comparison_1.endsWith; } });
Object.defineProperty(exports, "eq", { enumerable: true, get: function () { return comparison_1.eq; } });
Object.defineProperty(exports, "gt", { enumerable: true, get: function () { return comparison_1.gt; } });
Object.defineProperty(exports, "gte", { enumerable: true, get: function () { return comparison_1.gte; } });
Object.defineProperty(exports, "in", { enumerable: true, get: function () { return comparison_1.inOp; } });
Object.defineProperty(exports, "isNormalized", { enumerable: true, get: function () { return comparison_1.isNormalized; } });
Object.defineProperty(exports, "isNotNormalized", { enumerable: true, get: function () { return comparison_1.isNotNormalized; } });
Object.defineProperty(exports, "isNotNull", { enumerable: true, get: function () { return comparison_1.isNotNull; } });
Object.defineProperty(exports, "isNull", { enumerable: true, get: function () { return comparison_1.isNull; } });
Object.defineProperty(exports, "lt", { enumerable: true, get: function () { return comparison_1.lt; } });
Object.defineProperty(exports, "lte", { enumerable: true, get: function () { return comparison_1.lte; } });
Object.defineProperty(exports, "matches", { enumerable: true, get: function () { return comparison_1.matches; } });
Object.defineProperty(exports, "neq", { enumerable: true, get: function () { return comparison_1.neq; } });
Object.defineProperty(exports, "startsWith", { enumerable: true, get: function () { return comparison_1.startsWith; } });
var concat_1 = require("./expressions/operations/concat");
Object.defineProperty(exports, "concat", { enumerable: true, get: function () { return concat_1.concat; } });
Object.defineProperty(exports, "ConcatOp", { enumerable: true, get: function () { return concat_1.ConcatOp; } });
var math_1 = require("./expressions/operations/math");
Object.defineProperty(exports, "divide", { enumerable: true, get: function () { return math_1.divide; } });
Object.defineProperty(exports, "minus", { enumerable: true, get: function () { return math_1.minus; } });
Object.defineProperty(exports, "mod", { enumerable: true, get: function () { return math_1.mod; } });
Object.defineProperty(exports, "multiply", { enumerable: true, get: function () { return math_1.multiply; } });
Object.defineProperty(exports, "plus", { enumerable: true, get: function () { return math_1.plus; } });
Object.defineProperty(exports, "pow", { enumerable: true, get: function () { return math_1.pow; } });
// --Functions
var CypherFunctions_1 = require("./expressions/functions/CypherFunctions");
Object.defineProperty(exports, "Function", { enumerable: true, get: function () { return CypherFunctions_1.CypherFunction; } });
var aggregation_1 = require("./expressions/functions/aggregation");
Object.defineProperty(exports, "avg", { enumerable: true, get: function () { return aggregation_1.avg; } });
Object.defineProperty(exports, "collect", { enumerable: true, get: function () { return aggregation_1.collect; } });
Object.defineProperty(exports, "count", { enumerable: true, get: function () { return aggregation_1.count; } });
Object.defineProperty(exports, "max", { enumerable: true, get: function () { return aggregation_1.max; } });
Object.defineProperty(exports, "min", { enumerable: true, get: function () { return aggregation_1.min; } });
Object.defineProperty(exports, "percentileCont", { enumerable: true, get: function () { return aggregation_1.percentileCont; } });
Object.defineProperty(exports, "percentileDisc", { enumerable: true, get: function () { return aggregation_1.percentileDisc; } });
Object.defineProperty(exports, "stDev", { enumerable: true, get: function () { return aggregation_1.stDev; } });
Object.defineProperty(exports, "stDevP", { enumerable: true, get: function () { return aggregation_1.stDevP; } });
Object.defineProperty(exports, "sum", { enumerable: true, get: function () { return aggregation_1.sum; } });
exports.graph = __importStar(require("./expressions/functions/graph"));
__exportStar(require("./expressions/functions/list"), exports);
var load_csv_1 = require("./expressions/functions/load-csv");
Object.defineProperty(exports, "file", { enumerable: true, get: function () { return load_csv_1.file; } });
Object.defineProperty(exports, "linenumber", { enumerable: true, get: function () { return load_csv_1.linenumber; } });
var math_2 = require("./expressions/functions/math");
Object.defineProperty(exports, "abs", { enumerable: true, get: function () { return math_2.abs; } });
Object.defineProperty(exports, "acos", { enumerable: true, get: function () { return math_2.acos; } });
Object.defineProperty(exports, "asin", { enumerable: true, get: function () { return math_2.asin; } });
Object.defineProperty(exports, "atan", { enumerable: true, get: function () { return math_2.atan; } });
Object.defineProperty(exports, "atan2", { enumerable: true, get: function () { return math_2.atan2; } });
Object.defineProperty(exports, "ceil", { enumerable: true, get: function () { return math_2.ceil; } });
Object.defineProperty(exports, "cos", { enumerable: true, get: function () { return math_2.cos; } });
Object.defineProperty(exports, "cot", { enumerable: true, get: function () { return math_2.cot; } });
Object.defineProperty(exports, "degrees", { enumerable: true, get: function () { return math_2.degrees; } });
Object.defineProperty(exports, "e", { enumerable: true, get: function () { return math_2.e; } });
Object.defineProperty(exports, "exp", { enumerable: true, get: function () { return math_2.exp; } });
Object.defineProperty(exports, "floor", { enumerable: true, get: function () { return math_2.floor; } });
Object.defineProperty(exports, "haversin", { enumerable: true, get: function () { return math_2.haversin; } });
Object.defineProperty(exports, "isNaN", { enumerable: true, get: function () { return math_2.cypherIsNaN; } });
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return math_2.log; } });
Object.defineProperty(exports, "log10", { enumerable: true, get: function () { return math_2.log10; } });
Object.defineProperty(exports, "pi", { enumerable: true, get: function () { return math_2.pi; } });
Object.defineProperty(exports, "radians", { enumerable: true, get: function () { return math_2.radians; } });
Object.defineProperty(exports, "rand", { enumerable: true, get: function () { return math_2.rand; } });
Object.defineProperty(exports, "round", { enumerable: true, get: function () { return math_2.round; } });
Object.defineProperty(exports, "sign", { enumerable: true, get: function () { return math_2.sign; } });
Object.defineProperty(exports, "sin", { enumerable: true, get: function () { return math_2.sin; } });
Object.defineProperty(exports, "sqrt", { enumerable: true, get: function () { return math_2.sqrt; } });
Object.defineProperty(exports, "tan", { enumerable: true, get: function () { return math_2.tan; } });
__exportStar(require("./expressions/functions/path"), exports);
var predicate_1 = require("./expressions/functions/predicate");
Object.defineProperty(exports, "all", { enumerable: true, get: function () { return predicate_1.all; } });
Object.defineProperty(exports, "any", { enumerable: true, get: function () { return predicate_1.any; } });
Object.defineProperty(exports, "exists", { enumerable: true, get: function () { return predicate_1.exists; } });
Object.defineProperty(exports, "isEmpty", { enumerable: true, get: function () { return predicate_1.isEmpty; } });
Object.defineProperty(exports, "none", { enumerable: true, get: function () { return predicate_1.none; } });
Object.defineProperty(exports, "single", { enumerable: true, get: function () { return predicate_1.single; } });
__exportStar(require("./expressions/functions/scalar"), exports);
__exportStar(require("./expressions/functions/spatial"), exports);
__exportStar(require("./expressions/functions/string"), exports);
var temporal_1 = require("./expressions/functions/temporal");
Object.defineProperty(exports, "date", { enumerable: true, get: function () { return temporal_1.cypherDate; } });
Object.defineProperty(exports, "datetime", { enumerable: true, get: function () { return temporal_1.cypherDatetime; } });
Object.defineProperty(exports, "duration", { enumerable: true, get: function () { return temporal_1.duration; } });
Object.defineProperty(exports, "localdatetime", { enumerable: true, get: function () { return temporal_1.cypherLocalDatetime; } });
Object.defineProperty(exports, "localtime", { enumerable: true, get: function () { return temporal_1.cypherLocalTime; } });
Object.defineProperty(exports, "time", { enumerable: true, get: function () { return temporal_1.cypherTime; } });
// Procedures
var CypherProcedure_1 = require("./procedures/CypherProcedure");
Object.defineProperty(exports, "Procedure", { enumerable: true, get: function () { return CypherProcedure_1.CypherProcedure; } });
Object.defineProperty(exports, "VoidProcedure", { enumerable: true, get: function () { return CypherProcedure_1.VoidCypherProcedure; } });
/**
 * Utility functions
 * @group Utils
 * @hideGroups
 */
exports.utils = __importStar(require("./utils/utils"));
