// @neo4j/graphql/dist/schema/resolvers/composition/utils/get-authorization-context.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAuthorizationContext = getAuthorizationContext;
const cypher_builder_1 = __importDefault(require("@neo4j/cypher-builder"));
const debug_1 = __importDefault(require("debug"));
const constants_1 = require("../../../../constants");
const debug_object_1 = require("../../../../debug/debug-object");
const debug = (0, debug_1.default)(constants_1.DEBUG_AUTH);
const unauthorizedContext = {
    isAuthenticated: false,
    jwtParam: new cypher_builder_1.default.NamedParam("jwt", {}),
    isAuthenticatedParam: new cypher_builder_1.default.NamedParam("isAuthenticated", false),
};
async function getAuthorizationContext(context, authorization, jwtClaimsMap) {
    if (context.jwt) {
        const isAuthenticated = true;
        const jwt = context.jwt;
        (0, debug_object_1.debugObject)(debug, "using JWT provided in context", jwt);
        return {
            isAuthenticated,
            jwt,
            jwtParam: new cypher_builder_1.default.NamedParam("jwt", jwt),
            isAuthenticatedParam: new cypher_builder_1.default.NamedParam("isAuthenticated", isAuthenticated),
        };
    }
    if (!authorization) {
        debug("authorization settings not specified, request not authenticated");
        return unauthorizedContext;
    }
    if (context.token) {
        const jwt = await authorization.decode(context);
        if (jwt) {
            context.jwt = jwt;
            const isAuthenticated = true;
            (0, debug_object_1.debugObject)(debug, "successfully decoded JWT", jwt);
            return {
                isAuthenticated,
                jwt: context.jwt,
                jwtParam: new cypher_builder_1.default.NamedParam("jwt", context.jwt),
                isAuthenticatedParam: new cypher_builder_1.default.NamedParam("isAuthenticated", isAuthenticated),
                claims: jwtClaimsMap,
            };
        }
    }
    debug("request not authenticated");
    return unauthorizedContext;
}
